### PR TITLE
Automatically restart P2P shuffles when output worker leaves

### LINKED ISSUE #7353: Automatically restart P2P shuffles that were aborted due to leaving workers
With https://github.com/dask/distributed/pull/7326 we guarantee that a shuffle consistently fails if a *participating* worker dies (input-only workers will also trigger this since we currently do not guarantee exactly-once guarantees, see https://github.com/dask/distributed/issues/7324). This is currently a hard requirement since input tasks split their data by output partition and push these shards to the designated output workers.
If one of the output worker dies we loose the data that was already pushed to it. The tasks assigned to the dead worker at the time are not representative of the lost data and it is not sufficient to reschedule the lost tasks but we need to reschedule all of them.

Whenever a shuffle is rescheduled, we will generate new metadata on the scheduler side with a newly calculated output worker mapping. We will increase the `attempt` or `generation` counter on this metadata to distinguish this run from the earlier one. All transfer tasks need to be rescheduled and will need to execute at least once using the new metadata.

Rescheduling is not an atomic operation and we need to assume that old generation tasks are still running or are about to run while we are resetting the metadata on the scheduler, are broadcasting it to all workers and are transitioning the tasks.
Neither scheduler, nor workers do have a notion of 'attempts' or 'generations'. Therefore any finished task from an earlier generation, both failure (all tasks) and success (transfer+barrier), will corrupt an ongoing shuffle unless we can detect old-generation task-finished responses and handle them accordingly.

Summarized, we need to

1. Ensure that every task of a given generation is executed at least once
2. Suppress any finished tasks from an earlier generation

Iff we do not guaranteed deduplication on receiver end (https://github.com/dask/distributed/issues/7324) we further need to ensure that shards of an earlier generation must not be accepted by `Shuffle` instances of a newer generation. Otherwise, if the output_worker mappings between old and new generation map some output partitions to the same worker, old generation transfer tasks may already send some shards to the new generation receivers. Once this task is finished and rescheduled, this would effectively cause data duplication.
Apart from https://github.com/dask/distributed/issues/7324 we need weaker guarentees to avoid this and it is sufficient if we reject transfers that do not belong to the same generation.

3. Ensure transfers of an earlier generation do not contribute to the new generation, i.e. new generation receivers must reject old generation tranfers

---

1.) is trivially achieved by simply releasing all transfer tasks and the barrier task.
Note that not necessarily all unpack tasks will automatically be rescheduled. If the unpack tasks were already released beforehand and the dependents are still in memory, they will not be recomputed. This will affect some internal logic that counts how many output tasks where already processed to determine whether a shuffle is "done".

2.) The only possibility to entirely ignore old generation results is to expand the scheduler transition engine with this concept. Every Scheduler->Worker compute-task message will sign this request with a unique identifier (could be a simple counter / the stimulus id / etc.) and remembers the most recent identifier. The Worker will store this unique identifier in a TaskState attribute and will return it again in it's task-finished message. The scheduler handlers can then properly deduplicate the requests by ignoring stale responses.
This concept has already been successfully implemented in work stealing.
This would be the first modification of the actual scheduler, outside of an extension to enable P2P. I believe this additional guarantee would not be harmful and may even avoid some fringe races that are currently dealt with by sophisticated transitions.

Note: A mere transition hook is not sufficient since a task failure would already generate a client message indicating a failure. the transition hooks do not allow us to intercept messages to clients (nor workers) and I don't think this would be a table API

3.) If the replicated shuffle metadata includes an attempt/generation counter which is included in the shards submission. the receiving end will ignore all pushes that do not match it's own. If the sender is old generation, the sender task can simply err/finish (pending 1.)). If the sender is new generation it needs to repeat until the receiving end is updated.

cc @hendrikmakait @mrocklin

### PR DESCRIPTION
**Pull request by hendrikmakait**
Closes #7353 

~Blocked by and includes #7967~
~Blocked by and includes #7979~
~Blocked by and includes #7981~
~Blocked by and includes #7974~
- [ ] Tests added / passed
- [x] Passes `pre-commit run --all-files`

### CODE REVIEW

File: distributed/scheduler.py
Comment (hendrikmakait) [inline distributed/scheduler.py:1993]: Superseded by #7974

File: distributed/scheduler.py
Comment (hendrikmakait) [inline distributed/scheduler.py:5089]: Superseded by #7974

**Comment by hendrikmakait** [comment]
cc @wence-

**Review by wence- — COMMENTED** [review]
I wonder if some of the internal state management around the known shuffles can be refactored to make it clearer that invariants are being maintained.

File: distributed/scheduler.py
Comment (hendrikmakait) [reply distributed/scheduler.py:5746]: Good catch, this is already part of https://github.com/dask/distributed/pull/7974 but this branch hasn't been updated yet.

File: distributed/shuffle/_scheduler_plugin.py
Comment (hendrikmakait) [reply distributed/shuffle/_scheduler_plugin.py:370]: I failed to make this clear enough in the docstring: Due to the order of operations in `Scheduler.remove_worker`, it can happen that a barrier task will already have been transitioned to `released`, e.g., if it was `processing` on the removed worker. This would cause its shuffle to get archived before we reach this point. Since this archival was caused by the worker removal, we would like to recover from it and restart the shuffle, so we check for shuffles that were archived by the same stimulus.

File: distributed/shuffle/_scheduler_plugin.py
Comment (hendrikmakait) [reply distributed/shuffle/_scheduler_plugin.py:417]: On a long-running cluster, we would keep adding new entries to the dictionary and unnecessarily increase its size.

File: distributed/shuffle/_scheduler_plugin.py
Comment (hendrikmakait) [reply distributed/shuffle/_scheduler_plugin.py:111]: I'm certain there is, but I would like to keep the refactoring separate from actual logic changes. Let's keep this for a follow-up PR? (Would you be interested in taking that on?)

Comment (wence-) [reply distributed/shuffle/_scheduler_plugin.py:111]: Please open an issue and assign to me, and I will give it a go, thanks.

File: distributed/shuffle/_scheduler_plugin.py
Comment (hendrikmakait) [reply distributed/shuffle/_scheduler_plugin.py:386]: Good catch, addressed.

File: distributed/shuffle/_scheduler_plugin.py
Comment (hendrikmakait) [reply distributed/shuffle/_scheduler_plugin.py:413]: Individual `stimulus_id`s should be unique (think of them as a poor man's version of a span ID in tracing). They're usually created by combining a stimulus name (from a finite set of names) with a timestamp. For example: https://github.com/dask/distributed/blob/b7e5f8f97ef0eb95368122f29ac8915ae692f94e/distributed/scheduler.py#L5688-L5690

Comment (wence-) [reply distributed/shuffle/_scheduler_plugin.py:413]: Ah, ok.

File: distributed/shuffle/_buffer.py
Comment (hendrikmakait) [reply distributed/shuffle/_buffer.py]: This was a leftover from an intermediate state of this PR. While it technically doesn't hurt, I've removed it since it's unnecessary.

File: distributed/shuffle/_scheduler_plugin.py
Comment (hendrikmakait) [reply distributed/shuffle/_scheduler_plugin.py:111]: Will do once this is merged 👍

Comment (hendrikmakait) [reply distributed/shuffle/_scheduler_plugin.py:111]: https://github.com/dask/distributed/issues/8018

**Review by phofl — COMMENTED** [review]
Generally looks good to me, but not familiar enough with the code-base to be the final reviewer

File: distributed/shuffle/_scheduler_plugin.py
Comment (fjetter) [inline distributed/shuffle/_scheduler_plugin.py:55]: This appears to be not covered by tests. Why is this necessary then?

File: distributed/shuffle/_shuffle.py
Comment (fjetter) [inline distributed/shuffle/_shuffle.py:98]: This lack of coverage make me nervous. I think around the barrier there are various races we should test.

File: distributed/shuffle/_scheduler_plugin.py
Comment (fjetter) [inline distributed/shuffle/_scheduler_plugin.py:407]: IIUC this entire logic is just there to clean up. Behavior would not be impacted if we didn't do any of this, correct?

File: distributed/shuffle/_scheduler_plugin.py
Comment (fjetter) [inline distributed/shuffle/_scheduler_plugin.py:338]: why don't you want coverage to detect this? Seems like an important case

File: distributed/shuffle/_scheduler_plugin.py
Comment (fjetter) [inline distributed/shuffle/_scheduler_plugin.py:343]: same here

File: distributed/shuffle/_worker_plugin.py
Comment (fjetter) [inline distributed/shuffle/_worker_plugin.py:880]: I don't insist on this but I don't like these sleep patterns.

```python
def __init__(...):
    self._runs_condition = asyncio.Condition()

async def _close_shuffle_run(self, shuffle: ShuffleRun) -> None:
    await shuffle.close()
    async with self._runs_condition:
        self._runs.remove(shuffle)
        self._runs_condition.notify_all()

async def teardown(self, worker: Worker) -> None:
    ...
    async with self._runs_condition:
        await self._runs_condition.wait_for(lambda: not self._runs)

```

would be a clean alternative. Many people consider Conditions too complex but what I like about them is that they make this relationship very clear (and they unblock **immediately** which is nice for testing and such things).

As I said, I don't insist on this

Comment (fjetter) [reply distributed/shuffle/_worker_plugin.py:880]: To a more serious question: Is it possible for _runs to be repopulated at this point or are we locking everything up properly for this to not happen once we reach this point?

File: distributed/shuffle/tests/test_shuffle.py
Comment (fjetter) [inline distributed/shuffle/tests/test_shuffle.py:321]: does this mean that P2P is now retried `allowed-failures` times if a worker goes OOM?

Comment (fjetter) [reply distributed/shuffle/tests/test_shuffle.py:321]: Wouldn't be a dealbreaker but I also don't think this is useful. It's very unlikely that another P2P run attempt would be more successful.

However, there are of course also cases like spot interruption where this matters... Never mind!

File: distributed/shuffle/tests/test_shuffle.py
Comment (fjetter) [inline distributed/shuffle/tests/test_shuffle.py:703]: I guess this is out of scope for this PR but I think it would make sense to have an API to easily get access to the actual shuffle instnaces held by the plugins and have a `stage` attribute that indicates whether we're in transfer, barrier or unpack stage.
I would find this kind of verification nicer than waiting for task states.

**Review by fjetter — COMMENTED** [review]
Most of my comments are nits and you can ignore or use them.

The one thing I'm concerned about is that the coverage report indicates some rather critical code sections are not covered. We should look into that before merging

**Review by fjetter — COMMENTED** [review]
Most of my comments are nits and you can ignore or use them.

The one thing I'm concerned about is that the coverage report indicates some rather critical code sections are not covered. We should look into that before merging

File: distributed/shuffle/_scheduler_plugin.py
Comment (hendrikmakait) [reply distributed/shuffle/_scheduler_plugin.py:407]: Yes, this is state cleanup on the scheduler plugin.

File: distributed/shuffle/_scheduler_plugin.py
Comment (hendrikmakait) [reply distributed/shuffle/_scheduler_plugin.py]: It had some use in an earlier iteration. Removed.

File: distributed/shuffle/_scheduler_plugin.py
Comment (hendrikmakait) [reply distributed/shuffle/_scheduler_plugin.py:335]: https://github.com/dask/distributed/pull/7970#discussion_r1263396997

File: distributed/shuffle/_scheduler_plugin.py
Comment (hendrikmakait) [reply distributed/shuffle/_scheduler_plugin.py:340]: https://github.com/dask/distributed/pull/7970#discussion_r1263396997

File: distributed/shuffle/tests/test_shuffle.py
Comment (hendrikmakait) [reply distributed/shuffle/tests/test_shuffle.py:321]: > does this mean that P2P is now retried allowed-failures times if a worker goes OOM?
Yes, as there might be other causes apart from an output partition being too large.

File: distributed/shuffle/_scheduler_plugin.py
Comment (hendrikmakait) [reply distributed/shuffle/_scheduler_plugin.py]: Fair point, this bit has gotten out of hand.

File: distributed/shuffle/_scheduler_plugin.py
Comment (hendrikmakait) [reply distributed/shuffle/_scheduler_plugin.py:337]: Added a comment to explain

File: distributed/shuffle/_scheduler_plugin.py
Comment (hendrikmakait) [reply distributed/shuffle/_scheduler_plugin.py:344]: Added a comment to explain

File: distributed/shuffle/_worker_plugin.py
Comment (wence-) [reply distributed/shuffle/_worker_plugin.py]: `_runs` is added to in `_refresh_shuffle` which doesn't have a lock associated with it. But I am not sure if that can be running simultaneously with `teardown`.

File: distributed/shuffle/_worker_plugin.py
Comment (hendrikmakait) [inline distributed/shuffle/_worker_plugin.py:889]: Cosmetical refactoring to make it easier to understand whether we could potentially encounter races.

File: distributed/shuffle/_scheduler_plugin.py
Comment (wence-) [reply distributed/shuffle/_scheduler_plugin.py:337]: This seems like an ideal case for an `assert False, "Invariant broken"` ?

File: distributed/shuffle/_scheduler_plugin.py
Comment (wence-) [reply distributed/shuffle/_scheduler_plugin.py:344]: Similarly here.

**Review by wence- — COMMENTED** [review]
Not sure about the maintenance of invariants in `remove_worker`, plus a similar request to @fjetter for some coverage on edge cases?

File: distributed/shuffle/_scheduler_plugin.py
Comment (hendrikmakait) [reply distributed/shuffle/_scheduler_plugin.py]: > I suppose this is because the run_id is a unique token that defines the shuffle state object.

That's why I initially had a new `__eq__`. Then `__hash__` had to match it. Now I'm only using `__hash__`, so I _think_ there's no need for a custom `__eq__` that could potentially get outdated.

File: distributed/shuffle/_scheduler_plugin.py
Comment (hendrikmakait) [reply distributed/shuffle/_scheduler_plugin.py:337]: That would also work. I'm wondering if `assert False` is the right thing to add here given that `PYTHONOPTIMIZE` will strip them. It would work as an addition though.

File: distributed/shuffle/_worker_plugin.py
Comment (wence-) [inline distributed/shuffle/_worker_plugin.py:902]: This lock protects `_runs` wrt `_close_shuffle_run` but not wrt `_refresh_shuffle` I think.

File: distributed/shuffle/_worker_plugin.py
Comment (hendrikmakait) [reply distributed/shuffle/_worker_plugin.py]: `teardown` sets https://github.com/dask/distributed/blob/90eb9ea1d1991e2d1871dcbbe81f007dfbc8d821/distributed/shuffle/_worker_plugin.py#L890-L893

Once that is done, https://github.com/dask/distributed/blob/90eb9ea1d1991e2d1871dcbbe81f007dfbc8d821/distributed/shuffle/_worker_plugin.py#L810-L811 will raise. We don't yield the event loop in-between that raise and https://github.com/dask/distributed/blob/90eb9ea1d1991e2d1871dcbbe81f007dfbc8d821/distributed/shuffle/_worker_plugin.py#L832

File: distributed/shuffle/_worker_plugin.py
Comment (hendrikmakait) [reply distributed/shuffle/_worker_plugin.py]: I've renamed it to `_runs_cleanup_condition` to highlight that it's only concerned with cleanup. There's a different mechanism in place for adding to `self._runs`. (Feel free to refactor in a follow-up if you see a good way of doing so.)

File: distributed/shuffle/_scheduler_plugin.py
Comment (hendrikmakait) [reply distributed/shuffle/_scheduler_plugin.py:371]: Restarting a shuffle removes the `ShuffleState` from active states. The first `shuffle_transfer` task to call `shuffle_get_or_create` will cause the SchedulerPlugin to create a new `ShuffleState` with an incremented `run_id` and `_archived_by = None`.

File: distributed/shuffle/_scheduler_plugin.py
Comment (hendrikmakait) [reply distributed/shuffle/_scheduler_plugin.py:376]: I'm not 100% sure I'm following, but what _I think_ you're saying is a very good point.

File: distributed/shuffle/_scheduler_plugin.py
Comment (hendrikmakait) [reply distributed/shuffle/_scheduler_plugin.py:337]: raising a `RuntimeError` now

File: distributed/shuffle/_shuffle.py
Comment (hendrikmakait) [reply distributed/shuffle/_shuffle.py]: I haven't been able to come up with a scenario where this would be triggered, so I've removed it for now. If this ever pops up for somebody, I hope they'll send a bug report our way.

File: distributed/shuffle/_scheduler_plugin.py
Comment (hendrikmakait) [reply distributed/shuffle/_scheduler_plugin.py]: I haven't been able to come up with a scenario where this would happen, but given how messy worker shutdown can be, I'm not 100% certain this would _never_ happen. Left it in with a note for now.

File: distributed/shuffle/_scheduler_plugin.py
Comment (hendrikmakait) [reply distributed/shuffle/_scheduler_plugin.py]: refactored.

**Review by wence- — APPROVED** [review]
To the best of my understanding, this looks right!

**Pull request merged**
